File: crates/hyphae-server/tests/fixtures/sdk_v2_acceptance.py
from __future__ import annotations

import os
import threading
import time

from hyphae_sdk.v2 import CancellationToken, HyphaeClient, ProductError, RequestOptions


def fields(error: ProductError) -> dict[str, object]:
    return error.fields.__dict__


local = HyphaeClient.local(os.environ["HYPHAE_SOCKET"])
denied_local = HyphaeClient.local(
    os.environ["HYPHAE_SOCKET"], client_identity=os.environ["HYPHAE_DENIED_IDENTITY"]
)
http = HyphaeClient.http(os.environ["HYPHAE_ORIGIN"], bearer_token=os.environ["HYPHAE_TOKEN"])
denied_http = HyphaeClient.http(os.environ["HYPHAE_ORIGIN"])

cases = [
    ("sql_invalid_syntax", lambda client, options: client.sql("SELEC bad", options=options)),
    ("catalog_object_not_found", lambda client, options: client.catalog_object(999, options=options)),
    ("limit_exceeded", lambda client, options: client.sql("SELECT id FROM proof_items", options=options)),
]
for offset, (code, call) in enumerate(cases):
    request_id = 20_100 + offset
    limits = dict(RequestOptions().limits)
    if code == "limit_exceeded":
        limits["max_request_bytes"] = 1
    options = RequestOptions(request_id=request_id, limits=limits)
    errors = []
    for client in (local, http):
        try:
            call(client, options)
            raise AssertionError(f"{code} accepted")
        except ProductError as error:
            errors.append(fields(error))
    assert errors[0] == errors[1]
    assert errors[0]["code"] == code and errors[0]["request_id"] == request_id

expired_errors = []
for client in (local, http):
    expired = RequestOptions(
        request_id=20_110, deadline_micros=time.time_ns() // 1000 + 100
    )
    try:
        client.prove_sql("SELECT label FROM proof_items WHERE id = ?", [7], options=expired)
        raise AssertionError("expired request accepted")
    except ProductError as error:
        expired_errors.append(fields(error))
assert expired_errors[0] == expired_errors[1]
assert expired_errors[0]["code"] == "deadline_exceeded"

cancelled_errors = []
for transport, client in (("local", local), ("http", http)):
    token = CancellationToken()
    if transport == "local":
        threading.Timer(0.0001, token.cancel).start()
    else:
        token.cancel()
    cancelled = RequestOptions(request_id=20_111, cancellation=token)
    try:
        client.prove_sql("SELECT label FROM proof_items WHERE id = ?", [7], options=cancelled)
        raise AssertionError("cancelled request accepted")
    except ProductError as error:
        cancelled_errors.append(fields(error))
assert cancelled_errors[0] == cancelled_errors[1]
assert cancelled_errors[0]["code"] == "cancelled"

authorization_errors = []
for client in (denied_local, denied_http):
    try:
        client.structure_get(b"denied", options=RequestOptions(request_id=20_112))
        raise AssertionError("unauthorized request accepted")
    except ProductError as error:
        authorization_errors.append(fields(error))
assert authorization_errors[0] == authorization_errors[1]
assert authorization_errors[0]["code"] == "authorization_denied"

proven = http.prove_sql(
    "SELECT label FROM proof_items WHERE id = ?", [7], options=RequestOptions(request_id=20_120)
)
assert proven.kind == "proven"
assert proven.value["proof"].startswith(b"HYNPRF02")
assert proven.value["witness"].startswith(b"HYNWIT02")
proof_local = HyphaeClient.local(os.environ["HYPHAE_SOCKET"])
verified = proof_local.verify_proof(
    proven.value["proof"], proven.value["witness"], proven.value["trusted_anchor"],
    options=RequestOptions(request_id=20_121),
)
assert verified.kind == "proof_verification"
assert verified.value["semantic_reexecution_performed"]
proof_local.close()

with open(os.environ["HYPHAE_PYTHON_ARTIFACT"], "wb") as artifact:
    for value in (proven.value["proof"], proven.value["witness"], proven.value["trusted_anchor"]):
        artifact.write(len(value).to_bytes(8, "little"))
        artifact.write(value)

proof_local = HyphaeClient.local(os.environ["HYPHAE_SOCKET"])
proven = proof_local.prove_sql(
    "SELECT label FROM proof_items WHERE id = ?", [7], options=RequestOptions(request_id=20_122)
)
assert proven.value["proof"].startswith(b"HYNPRF02")
assert proven.value["witness"].startswith(b"HYNWIT02")
verified = http.verify_proof(
    proven.value["proof"], proven.value["witness"], proven.value["trusted_anchor"],
    options=RequestOptions(request_id=20_123),
)
assert verified.kind == "proof_verification"
assert verified.value["semantic_reexecution_performed"]
proof_local.close()

for offset, client in enumerate((local, http)):
    row_id = 20_200 + offset
    begun = client.transaction_begin(options=RequestOptions(request_id=20_200 + offset * 10))
    assert begun.kind == "explicit_transaction_status"
    assert begun.value["state"] == "active"
    handle = begun.value["handle"]
    staged = client.transaction_stage_sql(
        handle,
        "INSERT INTO proof_items (id, label) VALUES (?, ?)",
        [row_id, f"python-{offset}"],
        options=RequestOptions(request_id=20_201 + offset * 10),
    )
    assert staged.kind == "transaction_staged"
    assert staged.value["result"]["kind"] == "sql"
    rolled_back = client.transaction_rollback(
        handle, options=RequestOptions(request_id=20_202 + offset * 10)
    )
    assert rolled_back.kind == "transaction_rolled_back"
    selected = client.sql(
        "SELECT label FROM proof_items WHERE id = ?",
        [row_id],
        options=RequestOptions(request_id=20_203 + offset * 10),
    )
    assert selected.value["result"]["rows"] == []

local.close()
denied_local.close()

File: sdks/typescript/test/sdk-v2-acceptance.mjs
import assert from "node:assert/strict";
import { writeFile } from "node:fs/promises";

import { DEFAULT_LIMITS, HyphaeClient, ProductError } from "../dist/v2/index.js";

const local = HyphaeClient.local(process.env.HYPHAE_SOCKET);
const deniedLocal = HyphaeClient.local(
  process.env.HYPHAE_SOCKET,
  undefined,
  process.env.HYPHAE_DENIED_IDENTITY,
);
const http = HyphaeClient.http(process.env.HYPHAE_ORIGIN, { bearerToken: process.env.HYPHAE_TOKEN });
const deniedHttp = HyphaeClient.http(process.env.HYPHAE_ORIGIN);

const fields = (error) => {
  assert(error instanceof ProductError);
  return error.fields;
};
const failures = [
  ["sql_invalid_syntax", (client, options) => client.sql("SELEC bad", [], options)],
  ["catalog_object_not_found", (client, options) => client.catalogObject(999n, options)],
  ["limit_exceeded", (client, options) => client.sql("SELECT id FROM proof_items", [], options)],
];
for (const [offset, [code, call]] of failures.entries()) {
  const requestId = 30_100n + BigInt(offset);
  const options = { requestId, limits: code === "limit_exceeded" ? { ...DEFAULT_LIMITS, maxRequestBytes: 1 } : DEFAULT_LIMITS };
  const errors = [];
  for (const client of [local, http]) {
    try {
      await call(client, options);
      throw new Error(`${code} accepted`);
    } catch (error) {
      errors.push(fields(error));
    }
  }
  assert.deepEqual(errors[0], errors[1]);
  assert.equal(errors[0].code, code);
  assert.equal(errors[0].requestId, requestId);
}

const expiredErrors = [];
for (const client of [local, http]) {
  try {
    await client.proveSql(
      "SELECT label FROM proof_items WHERE id = ?",
      [7n],
      {},
      { requestId: 30_110n, deadlineMicros: BigInt(Date.now()) * 1000n + 100n },
    );
    throw new Error("expired request accepted");
  } catch (error) {
    expiredErrors.push(fields(error));
  }
}
assert.deepEqual(expiredErrors[0], expiredErrors[1]);
assert.equal(expiredErrors[0].code, "deadline_exceeded");

const cancelledErrors = [];
for (const [transport, client] of [["local", local], ["http", http]]) {
  const controller = new AbortController();
  if (transport === "local") queueMicrotask(() => controller.abort());
  else controller.abort();
  try {
    await client.proveSql(
      "SELECT label FROM proof_items WHERE id = ?",
      [7n],
      {},
      { requestId: 30_111n, signal: controller.signal },
    );
    throw new Error("cancelled request accepted");
  } catch (error) {
    cancelledErrors.push(fields(error));
  }
}
assert.deepEqual(cancelledErrors[0], cancelledErrors[1]);
assert.equal(cancelledErrors[0].code, "cancelled");

const authorizationErrors = [];
for (const client of [deniedLocal, deniedHttp]) {
  try {
    await client.structureGet(new TextEncoder().encode("denied"), { requestId: 30_112n });
    throw new Error("unauthorized request accepted");
  } catch (error) {
    authorizationErrors.push(fields(error));
  }
}
assert.deepEqual(authorizationErrors[0], authorizationErrors[1]);
assert.equal(authorizationErrors[0].code, "authorization_denied");

const proven = await local.proveSql("SELECT label FROM proof_items WHERE id = ?", [7n], {}, { requestId: 30_120n });
assert.equal(proven.kind, "proven");
assert.equal(new TextDecoder().decode(proven.value.proof.slice(0, 8)), "HYNPRF02");
assert.equal(new TextDecoder().decode(proven.value.witness.slice(0, 8)), "HYNWIT02");
const verified = await http.verifyProof(
  proven.value.proof,
  proven.value.witness,
  proven.value.trustedAnchor,
  { requestId: 30_121n },
);
assert.equal(verified.kind, "proof_verification");
assert.equal(verified.value.semanticReexecutionPerformed, true);
{
  const values = [proven.value.proof, proven.value.witness, proven.value.trustedAnchor];
  const encoded = new Uint8Array(values.reduce((total, value) => total + 8 + value.byteLength, 0));
  const view = new DataView(encoded.buffer);
  let offset = 0;
  for (const value of values) {
    view.setBigUint64(offset, BigInt(value.byteLength), true);
    offset += 8;
    encoded.set(value, offset);
    offset += value.byteLength;
  }
  await writeFile(process.env.HYPHAE_TYPESCRIPT_ARTIFACT, encoded);
}

const httpProven = await http.proveSql("SELECT label FROM proof_items WHERE id = ?", [7n], {}, { requestId: 30_122n });
assert.equal(new TextDecoder().decode(httpProven.value.proof.slice(0, 8)), "HYNPRF02");
assert.equal(new TextDecoder().decode(httpProven.value.witness.slice(0, 8)), "HYNWIT02");
const proofLocal = HyphaeClient.local(process.env.HYPHAE_SOCKET);
const localVerified = await proofLocal.verifyProof(
  httpProven.value.proof,
  httpProven.value.witness,
  httpProven.value.trustedAnchor,
  { requestId: 30_123n },
);
assert.equal(localVerified.value.semanticReexecutionPerformed, true);
await proofLocal.close();

await local.close();
await deniedLocal.close();
